Ensure test environment is set up with test data and soft dependencies before running tests

jidicula opened the pull request:
## Checklist

- [x] I've given this PR a concise, self-descriptive, and meaningful title
- [x] I've linked relevant issues in the PR body
- [x] I've applied the relevant labels to this PR
- [x] I've added relevant tests for my contribution
- [x] I've updated the documentation and/or added correct docstrings
- [x] I've assigned a reviewer

<!--- For the title, please observe the following rules:
	- Provide a concise and self-descriptive title
	- Do not include the applicable issue number in the title, do it in the PR body
	- If the PR is not ready for review, convert it to a draft.
-->

## Description
<!-- describe what the PR is about. Explain the approach and possible drawbacks.It's ok to repeat some text from the related issue. -->

**Why this change was necessary**
We want to ensure that test data has been pulled into the test
environment before running tests.

**What this change does**
Executes the download_data() CLI inside a pytest hook that runs first
when the test suite session begins.

**Any side-effects?**
Creates a new temp directory and populates it with test data.

## Linked issues
<!-- If the PR fixes any issues, indicate it here with issue-closing keywords: e.g. Resolves #XX, Fixes #XX, Addresses #XX. Note that if you want multiple issues to be autoclosed on PR merge, you must use the issue-closing verb before each relevant issue: e.g. Resolves #1, Resolves #2 -->

Resolves #118 - Make sure the environment is properly setup and
binaries are present before running pytest

May want to refactor with an @setup decorator:
https://stackoverflow.com/a/12600154/6310633.

kousu commented on azure-pipelines.yml:
```diff
@@ -1,26 +1,32 @@
-#- script: |
+# Always do this step
+- script: |
     # Beware: you MUST put `quit` at the end; Matlab does not auto-quit.
     # And worse, our Matlab license only allows one process at a time (TODO: verify this),
     # so a forgotten `quit` will block all tests on all branches.
     # https://www.mathworks.com/matlabcentral/answers/523194-matlab-script-in-batch-from-unix-command-line
-#    matlab -nodisplay -nosplash -r "disp('Unit tests!'); quit"
-#  displayName: 'Unit Tests'
+    matlab -nodisplay -nosplash -r "disp('Unit tests!'); quit"
+  displayName: 'Unit Tests'
```
I'd make unit tests a separate pipeline, or at least a separate job in this pipeline, since tests should run on each push to a branch and PR, while updating the website should only happen on pushes to `master`.

kousu replied:
We can make separate pipelines with separate .yml files:

![Screenshot_2020-06-18_13-55-02](https://user-images.githubusercontent.com/987487/85055446-90f7d700-b16b-11ea-87f9-6782cc8802be.png)

po09i replied:
@kousu I was gonna set it up in the same pipeline but without conditions for the CI and adding a condition to the doc generation to only happen on merges/push to master. Is this good practice or do you still recommend seperating in different jobs? Because I believe the triggers are still going to be the same so the conditions are still going to be required.

po09i replied:
@kousu I think your way is more maintainable in the long run so I'l switch to that.

po09i replied:
@kousu I created a new pipeline, could you go approve it please?

jcohenadad commented:
@po09i good habit to close the branch after merging so we don't end up with stale branches (requires maintenance)

jidicula committed:
[test][conftest] Download test_data before running test suite

**Why this change was necessary**
We want to ensure that test data has been pulled into the test
environment before running tests.

**What this change does**
Executes the download_data() CLI inside a pytest hook that runs first
when the test suite session begins.

**Any side-effects?**
Creates a new temp directory and populates it with test data.

Should this temp directory be removed after the testing run is
complete? Can we expect that the testing environment will be connected
to the internet?

**Additional context/notes/links**

Resolves #118 - Make sure the environment is properly setup and
binaries are present before running pytest

May want to refactor with an @setup decorator:
https://stackoverflow.com/a/12600154/6310633.

jidicula commented:
Questions:

1. Should the temp directory be removed after the testing run is
complete? Can we expect that the testing environment will be connected
to the internet (so they would be able to redownload test data each time the testing suite is run)?

2. Not sure if this is the right way to go about it – I was mostly following the SCT approach for test data download but it still seems a bit janky to me.

Next steps:

- [x] check for dcm2niix with litmus command
- [x] check for prelude with litmus command
- [x] make those soft dependency checks optional, and add some conditions to the corresponding tests to only run if the user specifies
- [ ] ~~add an `--all` flag that runs pytest with all the flags for soft dependencies~~ vanilla `pytest` runs all tests, user must specify if they're testing without the soft dependencies. See RTD build for details.

jcohenadad commented:
> Should the temp directory be removed after the testing run is complete? 

I guess we could, and have a flag to keep it for debugging purpose

> Can we expect that the testing environment will be connected to the internet (so they would be able to redownload test data each time the testing suite is run)?

not sure i understand-- do you mean for the CI?

po09i commented:
Current implementation uses testing_data which is assumed to be in the root of shimming-toolbox. If this is changed lets keep in mind that the `.gitignore` will have to be changed and the `__dir_testing__` variable as well. Not an issue but something to put on the todo list.

Ideally testing-data is downloaded once and not downloaded again if testing-data is intact or if the data is up to data but it seems like quite a bit of work something minor. So I would go with what Julien said.

>  Can we expect that the testing environment will be connected to the internet (so they would be able to redownload test data each time the testing suite is run)?

I think this is a fair assumption, maybe a flag could be set up with a path to a local version of testing-data if the need ever arises

jidicula commented:
> not sure i understand-- do you mean for the CI?

I mean for users – are we assuming that the device they've installed shimming-toolbox on will be connected to the Internet to be able to fetch test data if they run `pytest` on it?

jidicula commented:
> I think this is a fair assumption, maybe a flag could be set up with a path to a local version of testing-data if the need ever arises

out of scope for now, but definitely a future ticket

jidicula committed:
[refactor][conftest] Use __dir_testing__ pkg variable

**Why this change was necessary**
The package already has a global available for the testing dir
location and dirname.

**What this change does**
See title.

**Any side-effects?**
None.

**Additional context/notes/links**

Related to https://github.com/shimming-toolbox/shimming-toolbox/pull/136#issuecomment-705301251

jidicula commented:
🎉 Excellent, I found [this](https://docs.pytest.org/en/latest/example/markers.html#custom-marker-and-command-line-option-to-control-test-runs) pytest feature that should give us the ability to selectively run tests with the `dcm2niix` and `prelude` soft dependencies – we can add a specific setup in `conftest.py` checking if those deps are available, then add the corresponding decorator to any test function that requires that soft dependency.

edit: jk we're going to use markers defined in a `pytest.ini` file, scroll to top of the same page for an example.

jidicula committed:
[style][tests] Reformat test scripts

jidicula committed:
[test][prelude,dcm2niix] Add pytest markers for soft dependencies

**Why this change was necessary**
We want to be able to conditionally run some tests for interfacing
shimming-toolbox with its soft dependencies.

Currently, `prelude` is the only tool with its functionality, but in
the future, users may wish to use a different tool covering the same
functionality. Therefore, tests requiring interfacing with `prelude`
don't need to be run if `prelude` is not installed, and this case must
be covered by the shimming-toolbox testing suite.

`dcm2niix` is only needed by the `dcm2bids` library, which itself is
only needed by users who don't already have their files in the BIDS
format. If they already have BIDS files, they don't need to install
`dcm2niix` and therefore the shimming-toolbox testing suite must cover
the case where dcm2niix is not installed.

**What this change does**
This change uses the pytest feature allowing the marking of test
functions with specific markers[1]. Marked test functions only run
when the corresponding marker is passed as an argument to the `-m`
option to pytest. This change also updates the accompanying usage
instructions for this feature.

**Any side-effects?**
The minimal `$ pytest` command still runs all test functions. You must
specifically opt out of tests depending on the soft dependencies. See
accompanying documentation for details.

**Rejected implementation options**
pytest provides the functionality for custom markers specified at
runtime[3]. This option was rejected as each custom environment is
only defined inline in the decorator, which could lead to divergence
of decorators during an incomplete refactor. The chosen implementation
allows configuration of markers in a single source of truth,
simplifying the refactor and addition of new markers.

The desired functionality could also be implemented using the
`pytest.mark.skipif` pattern and hooking that to an automated dependency
check. However, this creates implicit rather than explicit
behaviour and could result in duplicated condition definitions in each
decorator.

pytest also provides functionality for running only tests that match a
provided keyword expression[4], but this approach was rejected as it
relies on consistent naming conventions across existing tests
depending on either of the soft dependencies. Using this approach
would require a significant refactor of function and accompanying test
function names.

**Additional context/notes/links**

You can also mark whole modules with a `pytest.mark` decorator[2].

Note that the markers are defined in a SSOT in `pytest.ini`. New
markers can be defined in this file, following the existing
formatting. The decorators for test functions are in the format
`@pytest.mark.markername`.

See also:
1: https://docs.pytest.org/en/latest/example/markers.html
2:
https://docs.pytest.org/en/latest/example/markers.html#marking-whole-classes-or-modules
3:
https://docs.pytest.org/en/latest/example/markers.html#custom-marker-and-command-line-option-to-control-test-runs
4: https://docs.pytest.org/en/latest/usage.html#specifying-tests-selecting-tests

Resolves #118 - Make sure the environment is properly setup and binaries are present before running pytest

jidicula committed:
[test][conftest] Test existence of soft dependencies conditionally

**Why this change was necessary**
We only need to check the existence of soft dependencies when we're
testing for those dependencies. Therefore, these tests should be
decorated with the corresponding mark.

**What this change does**
Adds test functions in conftest to ensure that the soft dependencies
are installed and available to shimming-toolbox.

**Any side-effects?**
None.

**Additional context/notes/links**

Note that markers cannot be easily added to fixtures. I used a
workaround documented in pytest's issues[1].

Fixtures cannot be passed to a class signature, and it would
duplicate a lot of code to pass to every test class methods, so
instead we can use the `@pytest.mark.usefixtures()`
decorator[2]. Unfortunately, this cannot be used for multiple
functions within a test file unless it contains a test class[2.Warning].

1: https://github.com/pytest-dev/pytest/issues/1368#issuecomment-466339463
2: https://docs.pytest.org/en/stable/fixture.html#use-fixtures-in-classes-and-modules-with-usefixtures

Resolves #118 - Make sure the environment is properly setup and binaries are present before running pytest

jidicula committed:
[chore][conftest] Clean up WIP changes

**Why this change was necessary**
I left some WIP code and comments.

**What this change does**
See title.

**Any side-effects?**
New output message upon download completion.

**Additional context/notes/links**

Resolves #118 - Make sure the environment is properly setup and binaries are present before running pytest

jidicula commented:
Ok, I think this is ready for review. If you're reviewing it, I recommend you read my commit messages to understand some of my reasoning in taking this approach.

jidicula commented:
Note that once this is merged, the correct marker must be added to any tests depending on those soft dependencies. We can also add new markers for any future soft dependencies, or for other conditional tests.

jidicula committed:
[tests][conftest] Update header

jcohenadad commented on test/test_prelude.py:
```diff
@@ -85,10 +91,16 @@ def test_non_default_mask(self):
 
         # Call prelude with mask (is_unwrapping_in_2d is also used because it is significantly faster)
-        unwrapped_phase_e1 = prelude(self.phase_e1, self.mag_e1, self.affine_phase_e1, mask, is_unwrapping_in_2d=True)
+        unwrapped_phase_e1 = prelude(
```
@jidicula do you have an automatic formatting in your IDE? if so maybe we should discuss formatting conventions. I personally don't mind how the code was written before-- it was still compliant with PEP8. PEP8 is fairly flexible when it comes to how code should be wrapped, so it's a matter of internal convention (i tend to like concise code, not spread out across multiple lines)

jidicula replied:
Yes, sorry about that! I'll disable it in my local settings for this repo.

jidicula replied:
Shall I try to roll back these formatting changes in a fixup commit?

jcohenadad replied:
if you don't mind-- otherwise no big deal- it's still clean code 😊

jidicula committed:
Revert "[style][tests] Reformat test scripts"

This reverts commit 1b5cc4dd16986b90fd2f37ceda3747da93a2a7b3.

jidicula committed:
[fix][test_prelude] Restore missing import

po09i commented:
What is the new situation for the testing_data? I noticed the code uses `__dir_testing__` which leads to shimming-toolbox/testing_data so the new tmp/ directory won't be used.

Also if the testing_data is downloaded in the new tmp folder, we probably want to remove the call to `st_download_data` in .travis.yml

jidicula commented:
> What is the new situation for the testing_data? I noticed the code uses `__dir_testing__` which leads to shimming-toolbox/testing_data so the new tmp/ directory won't be used.
> 
> Also if the testing_data is downloaded in the new tmp folder, we probably want to remove the call to `st_download_data` in .travis.yml

There's not really a new situation for testing_data afaik – it's creating `./testing_data` and downloading the data into it. I think this is the expected location for the testing data, correct?

I'll update the Travis config as well.

jidicula committed:
[refactor][conftest] Clarify variable names to reflect purpose

**Why this change was necessary**
The previous names of variables for paths did not accurately reflect
their purposes.

**What this change does**
See title.

**Any side-effects?**
None

**Additional context/notes/links**

Resolves https://github.com/shimming-toolbox/shimming-toolbox/pull/136#issuecomment-706474241
Related to #118

jidicula committed:
[test][Travis] Remove test data download from Travis config

**Why this change was necessary**
Since the testing suite already downloads the test data, we can remove
this functionality from the Travis build setup.

**What this change does**
See title.

**Any side-effects?**
Removes redundancy of test data download.

**Additional context/notes/links**

Resolves https://github.com/shimming-toolbox/shimming-toolbox/pull/136#issuecomment-706474241
Related to #118

jidicula commented:
Coveralls [seems to be failing](https://travis-ci.com/github/shimming-toolbox/shimming-toolbox/builds/189211908#L463) in our Travis build, but it's after success. Does something have to be fixed for this?

Edit: seems related to [this](https://github.com/z4r/python-coveralls/issues/74#issuecomment-582170116)
